subsys/net/ip/tcp.c

```diff
@@ -1344,3 +1412,1160 @@ int net_tcp_parse_opts(struct net_pkt *pkt, int opt_totlen,
 	NET_ERR("Invalid TCP opt: %d len: %d", opt, optlen);
 	return -EINVAL;
 }
+
+int tcp_hdr_len(struct net_pkt *pkt)
+{
+	struct net_tcp_hdr hdr, *tcp_hdr;
+
+	tcp_hdr = net_tcp_get_hdr(pkt, &hdr);
+	if (tcp_hdr) {
+		return NET_TCP_HDR_LEN(tcp_hdr);
+	}
+
+	return 0;
+}
+
+int net_tcp_recv(struct net_context *context, net_context_recv_cb_t cb,
+		 void *user_data)
+{
+	NET_ASSERT(context->tcp);
+
+	if (context->tcp->flags & NET_TCP_IS_SHUTDOWN) {
+		return -ESHUTDOWN;
+	} else if (net_context_get_state(context) != NET_CONTEXT_CONNECTED) {
+		return -ENOTCONN;
+	}
+
+	context->recv_cb = cb;
+	context->tcp->recv_user_data = user_data;
+
+	return 0;
+}
+
+static void queue_fin(struct net_context *ctx)
+{
+	struct net_pkt *pkt = NULL;
+	int ret;
+
+	ret = net_tcp_prepare_segment(ctx->tcp, NET_TCP_FIN, NULL, 0,
+				      NULL, &ctx->remote, &pkt);
+	if (ret || !pkt) {
+		return;
+	}
+
+	ret = net_tcp_send_pkt(pkt);
+	if (ret < 0) {
+		net_pkt_unref(pkt);
+	}
+}
+
+int net_tcp_put(struct net_context *context)
+{
+	if (net_context_get_ip_proto(context) == IPPROTO_TCP) {
+		if ((net_context_get_state(context) == NET_CONTEXT_CONNECTED ||
+		     net_context_get_state(context) == NET_CONTEXT_LISTENING)
+		    && !context->tcp->fin_rcvd) {
+			NET_DBG("TCP connection in active close, not "
+				"disposing yet (waiting %dms)", FIN_TIMEOUT);
+			k_delayed_work_submit(&context->tcp->fin_timer,
+					      FIN_TIMEOUT);
+			queue_fin(context);
+			return 0;
+		}
+
+		return -ENOTCONN;
+	}
+
+	return -EOPNOTSUPP;
+}
+
+int net_tcp_listen(struct net_context *context)
+{
+	if (net_context_get_ip_proto(context) == IPPROTO_TCP) {
+		net_tcp_change_state(context->tcp, NET_TCP_LISTEN);
```

Comment:
Is `net_tcp_change_state()` exported? If so, can it be used only within `tcp.c`, and marked as static? Same idea for `net_tcp_get_state()`.